crates/accelerate/src/sparse_observable.rs

```diff
@@ -0,0 +1,1677 @@
+    /// string.
+    #[inline]
+    fn py_label(&self) -> &'static str {
+        match self {
+            Self::X => "X",
+            Self::Plus => "+",
+            Self::Minus => "-",
+            Self::Y => "Y",
+            Self::Right => "r",
+            Self::Left => "l",
+            Self::Z => "Z",
+            Self::Zero => "0",
+            Self::One => "1",
+        }
+    }
+
+    /// Get the name of this `BitTerm`, which is how Python space refers to the integer constant.
+    #[inline]
+    fn py_name(&self) -> &'static str {
+        match self {
+            Self::X => "X",
+            Self::Plus => "PLUS",
+            Self::Minus => "MINUS",
+            Self::Y => "Y",
+            Self::Right => "RIGHT",
+            Self::Left => "LEFT",
+            Self::Z => "Z",
+            Self::Zero => "ZERO",
+            Self::One => "ONE",
+        }
+    }
+
+    /// Attempt to convert a `u8` into `BitTerm`.
+    ///
+    /// Unlike the implementation of `TryFrom<u8>`, this allows `b'I'` as an alphabet letter,
+    /// returning `Ok(None)` for it.  All other letters outside the alphabet return the complete
+    /// error condition.
+    #[inline]
+    fn try_from_u8(value: u8) -> Result<Option<Self>, BitTermFromU8Error> {
+        match value {
+            b'+' => Ok(Some(BitTerm::Plus)),
+            b'-' => Ok(Some(BitTerm::Minus)),
+            b'0' => Ok(Some(BitTerm::Zero)),
+            b'1' => Ok(Some(BitTerm::One)),
+            b'I' => Ok(None),
+            b'X' => Ok(Some(BitTerm::X)),
+            b'Y' => Ok(Some(BitTerm::Y)),
+            b'Z' => Ok(Some(BitTerm::Z)),
+            b'l' => Ok(Some(BitTerm::Left)),
+            b'r' => Ok(Some(BitTerm::Right)),
+            _ => Err(BitTermFromU8Error(value)),
+        }
+    }
+}
+
+static BIT_TERM_PY_ENUM: GILOnceCell<Py<PyType>> = GILOnceCell::new();
+static BIT_TERM_INTO_PY: GILOnceCell<[Option<Py<PyAny>>; 16]> = GILOnceCell::new();
+
+/// Construct the Python-space `IntEnum` that represents the same values as the Rust-spce `BitTerm`.
+///
+/// We don't make `BitTerm` a direct `pyclass` because we want the behaviour of `IntEnum`, which
+/// specifically also makes its variants subclasses of the Python `int` type; we use a type-safe
+/// enum in Rust, but from Python space we expect people to (carefully) deal with the raw ints in
+/// Numpy arrays for efficiency.
+///
+/// The resulting class is attached to `SparseObservable` as a class attribute, and its
+/// `__qualname__` is set to reflect this.
+fn make_py_bit_term(py: Python) -> PyResult<Py<PyType>> {
+    let terms = [
+        BitTerm::X,
+        BitTerm::Plus,
+        BitTerm::Minus,
+        BitTerm::Y,
+        BitTerm::Right,
+        BitTerm::Left,
+        BitTerm::Z,
+        BitTerm::Zero,
+        BitTerm::One,
+    ]
+    .into_iter()
+    .flat_map(|term| {
+        let mut out = vec![(term.py_name(), term as u8)];
+        if term.py_name() != term.py_label() {
```

Comment:
Yeah, technically it makes `IntEnum.+` possible as well, but `Enum` subclasses know about this possibility - you can use `[]` syntax to get those, like `BitTerm["+"]`.  You can also do `getattr(BitTerm, "+")` and it'll work as well - Python class attributes really are just strings, and just because you can't spell it as legal syntax in the attribute getter doesn't mean it's not possible to access.

But the "canonical" form of them (and the only one I properly documented) is the all-caps names forms.

Reply:
Ok, I see 👍🏻 and the support of both `"+"` and `"PLUS"` is just a convenience, or is it necessary? (Not asking to change it, just to understand 🙂 )

Reply:
Just a convenience - I thought it might be useful to be able to do stuff like `[BitTerm[x] for x in "XrlYX+-01"]` potentially.